### PR TITLE
Add throttling

### PR DESCRIPTION
**Pull request by aocenas**

### CODE REVIEW

File: public/app/features/dashboard/state/runRequest.ts [hunk 1]
```diff
@@ -71,6 +82,28 @@ export function processResponsePacket(packet: DataQueryResponse, state: RunningQ
   return { packets, panelData };
 }
 
+function getCombinedState(states: LoadingState[]) {
```
Comment (aocenas) [inline public/app/features/dashboard/state/runRequest.ts]: @torkelo bit struggling with what the combined state should be. This works somehow as my assumption is we should take into account all requests and not tell the observer that we are Done where there are still some requests Loading.

The problem is though that with added throttling, if we pass on a PanelData after some time it will still have state Loading and panel won't show anything. I think it would make sense to keep the Loading state but to show what ever data we have so far and show longer loading series later on.

Does that make sense? If so I would try to change the logic of the panel to show even data when the overall state is Loading.

File: public/app/features/dashboard/state/runRequest.ts [hunk 2]
```diff
@@ -86,7 +119,12 @@ export function runRequest(datasource: DataSourceApi, request: DataQueryRequest)
       series: [],
       request: request,
     },
-    packets: {},
+    packets: {
```
Comment (aocenas) [inline public/app/features/dashboard/state/runRequest.ts]: Just for testing, this will keep the panel state as Loading indefinitely.

File: public/app/features/dashboard/state/PanelQueryRunner.ts [hunk 3]
```diff
@@ -158,7 +157,20 @@ export class PanelQueryRunner {
...
 
-      this.pipeToSubject(runRequest(ds, request));
+      const reqObservable = runRequest(ds, request);
+      this.pipeToSubject(
+        reqObservable.pipe(
+          throttle(
+            ev => {
+              return merge(
+                interval(1000),
+                reqObservable.pipe(filter((data: PanelData) => data.state === LoadingState.Done))
```
Comment (torkelo) [inline public/app/features/dashboard/state/PanelQueryRunner.ts]: this block of codes needs some comments :)  how can you use the same observable twice without share()?

Comment (aocenas) [reply public/app/features/dashboard/state/PanelQueryRunner.ts]: Definitely still need some polish and add comments but wanted another opinion on the issue with the combined state.

Here it works because `runRequest` already returns observable that is using `share()` but yeah that is totally not obvious here (and no special type for that either). Maybe using `share()` again here would make sense so we do not rely on internal implementation of `runRequest`.

Comment (aocenas) [reply public/app/features/dashboard/state/PanelQueryRunner.ts]: Ok actually this does not seem to work right now so need to investigate :(

File: public/app/features/dashboard/state/runRequest.ts (hunk 1 above)
Comment (torkelo) [reply public/app/features/dashboard/state/runRequest.ts]: Do we need to combine state? Can’t we just use last response state or Done?

Comment (aocenas) [reply public/app/features/dashboard/state/runRequest.ts]: If there is one packet that is still Loading and we get another one that is Done should we return state Done upstream? I assume this would remove any loading state in panel even though there are actually some data still loading which does not seem correct to me.

Comment (torkelo) [reply public/app/features/dashboard/state/runRequest.ts]: Think last packet state wins will work fine in 99.9% of cases

Comment (torkelo) [reply public/app/features/dashboard/state/runRequest.ts]: Loading state is currently set using last returned state (no matter what key on response) think keeping loading state per key will only complicate code here and complicate things for data sources.

Think of this way, 2 queries 
Query 1 finishes , datasource returns data with key A and state loading as there is still one query that has yet to complete (Something a datasource will have to track internally if it wants to) 
Query 2 finishes and datasource sends another packet with key B and state Done. 

runRequest will only need to care about latest state

Comment (aocenas) [reply public/app/features/dashboard/state/runRequest.ts]: Hmm not sure I get you there. If I do 2 requests:
```
req  1: Loading ----> Done
req  2: Loading -----------------> Done
result: Loading ----> Done ------> Done
```
Which seem wrong to me and something that is probably going to happen most of the time, isn't it?

Comment (aocenas) [reply public/app/features/dashboard/state/runRequest.ts]: > datasource returns data with key A and state loading as there is still one query that has yet to complete

Is this something that is already implemented? Because I do not see that in prometheus. It seems it just sets Done state on the response as soon as it returns no matter whether other queries are still in flight.

Comment (torkelo) [reply public/app/features/dashboard/state/runRequest.ts]: > Is this something that is already implemented? 

Not yet. But it should be. It's the only way I can see it working, as it's only the data source that knows the current loading state and it communicates that via the last emitted response. 

```
req  1: Loading ----> Done
req  2: Loading -----------------> Done
result: Loading ----> Done ------> Done
`` 

that is not how it works.

Req to datasource with 2 queries (which has an api that has no batching support like prometheus so has to issue multiple http requests) 
query A: http request sent -----  (emit result with key A, state: loading)
query B: http request sent ------------------------------- (emit result with key B state: Done)
result from runRequest: 
..
..
no response after 200ms emit loading result 
got result with key A, oh loading state, I guess more is coming
..
..
got result with key B, oh state done, I guess nothing more is comig. 
`` 

Basically I see state as a way for the data source to inform Grafana/runRequest what the CURRENT state should be (which is why it only needs to look at last emitted response). 
Not sure what the alternative would be in your world where state of old response packets are considererd. How would the data source use that?

Comment (aocenas) [reply public/app/features/dashboard/state/runRequest.ts]: I am still confused by:
> Not yet. But it should be.

Which I assume means my example is right in the sense that it describes how it currently works.

> that is not how it works.

So not sure if you mean it should not work like that in the future and are explaining how it should work conceptually.

> Not sure what the alternative would be in your world where state of old response packets is considered. How would the data source use that?

> Basically I see state as a way for the data source to inform Grafana/runRequest what the CURRENT state should be for the whole observable stream

I mean I think of it the same way I just implemented it here instead of in the datasource. So if this should be implemented in the datasource then cool (as I said I don't see this already implemented anywhere), but I think the logic would be kinda similar there and also duplicated in each datasource I assume.

So in general I see the datasource response as stream of streams (one per each target each with it's own state) which are just merged and we wan't to unify the state of those multiple streams into one state. Only thing that does not seem to be clear here is where that should be implemented, whether here or in each datasource.

Comment (torkelo) [reply public/app/features/dashboard/state/runRequest.ts]: > Which I assume means my example is right in the sense that it describes how it currently works.

No your example had bunch of "Loading" events that would never be sent (is not sent).

> So not sure if you mean it should not work like that in the future and are explaining how it should work conceptually.

I am explaining how the state property is supposed to be used by data sources in the DataQueryResponse. It's an optional property to describe the current state of the request. If not specified Done is assumed after first response.  

>  but I think the logic would be kinda similar there and also duplicated in each datasource I assume.

I think not. What you propose would mean a lot more added complexity to each data source and a lot of more emitted responses to update state old keys. 

> So in general I see the datasource response as stream of streams (one per each target each with it's own state) 

This is up the data source. Grafana will handle it as a single stream of responses and a key to  inform Grafana how to merge/combine responses. But there is only one state (panel has only one spinner/state indicator). 


One more detail to flesh out why your example I think adds more complexity (and is not how it how Prometheus is implemented).

```
req  1: Loading ----> Done
req  2: Loading -----------------> Done
result: Loading ----> Done ------> Done
```
There are no loading events emitted (currently). Only Done events. Think this simplifies things a lot in the data source. It can just emit events with the current of the whole request. Feel like emitting events from the data source for every start of request should not be needed and will only complicate things. runRequest handles the initial loading state by triggering a loading result after 200ms of no initial results.

Comment (torkelo) [reply public/app/features/dashboard/state/runRequest.ts]: I think I understand where your coming from because the previous solution PanelQueryState did deal with multiple states per stream. But with runRequest I saw opportunity to simplify a lot of this logic by pushing "managing multiple streams" down to the data source, which I think simplifies the logic for both.

**Review by aocenas — COMMENTED** [review]
@torkelo tried to fix the issue with the partial loading. Kept quite a bit of debug code and needs test but wanted to discuss this before that.

**Comment by torkelo** [comment]
Closing this as I think most is merged via other PRs, the remaining thing is to add throttling to PanelQueryRunner. 

Think for now a simple throttleTime(50) or similar low value should be fine.
